controllers/storageautoscaler/storageautoscaler_controller.go

```diff
@@ -0,0 +1,255 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// StorageAutoscalerReconciler is the reconciler for StorageAutoscaler objects
+type StorageAutoscalerReconciler struct {
+	client.Client
+	Log               logr.Logger
+	OperatorNamespace string
+	SyncMap           *sync.Map
+	EventCh           chan event.GenericEvent
+}
+
+// SetupWithManager sets up the reconciler with the manager
+func (r *StorageAutoscalerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// get the eventCh from the storage autoscaler scraper
+	storageAutoscalerReconcilerController := ctrl.NewControllerManagedBy(mgr).
+		For(&ocsv1.StorageAutoScaler{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		// watch for generic events to trigger the reconcile
+		WatchesRawSource(source.Channel(r.EventCh,
+			&handler.EnqueueRequestForObject{},
+		))
+
+	return storageAutoscalerReconcilerController.Complete(r)
+}
+
+// +kubebuilder:rbac:groups="monitoring.coreos.com",resources=prometheuses/api,resourceNames=k8s,verbs=get;create;update
+// +kubebuilder:rbac:groups="",resources=namespaces,verbs=get
+
+// Reconcile reconciles the StorageAutoscaler object
+func (r *StorageAutoscalerReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	r.Log.Info("reconciling storage autoscaler", "namespace", request.Namespace, "name", request.Name)
+
+	// list the storage autoscaler reconciler object
+	storageAutoScaler := &ocsv1.StorageAutoScaler{}
+	err := r.Get(ctx, client.ObjectKey{Namespace: request.Namespace, Name: request.Name}, storageAutoScaler)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("storage autoscaler not found for reconcile request", "namespace", request.Namespace, "name", request.Name)
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		r.Log.Error(err, "failed to get storage autoscaler")
+		return reconcile.Result{}, err
+	}
+
+	// if no expansion is in-progress update the phase to "not-started"
+	if storageAutoScaler.Status.Phase == "" {
+		originalStorageAutoScaler := storageAutoScaler.DeepCopy()
+		storageAutoScaler.Status.Phase = "NotStarted"
+		err := r.updateStatus(ctx, storageAutoScaler, originalStorageAutoScaler)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// if the expansion is InProgress or Failed, verify the scaling
+	if storageAutoScaler.Status.Phase == "InProgress" || storageAutoScaler.Status.Phase == "Failed" {
+		return r.verifyScaling(ctx, storageAutoScaler)
+	}
+
+	// if the last expansion is succeeded 10 minutes ago, do not scale
+	if storageAutoScaler.Status.LastExpansion != nil {
+		if time.Since(storageAutoScaler.Status.LastExpansion.CompletionTime.Time).Seconds() < 600 {
```

Comment:
If we are checking for the ceph health ok, I really don't see the need for this 10min delay.

Reply:
Ok sounds good, how about logging that it was rejected because of that 10 min wait?